akka-contrib/src/main/scala/akka/contrib/pattern/ReceivePipeline.scala

```diff
@@ -2,46 +2,113 @@ package akka.contrib.pattern
+   * Interceptor return value that indicates that the message has been handled
+   * completely. The message will not be passed to inner interceptors
+   * (or to the advised actor's receive).
+   */
+  case object HandledCompletely extends InterceptorResult
+
+  private def withDefault(interceptor: Interceptor): Interceptor = interceptor.orElse({ case msg => Inner(msg) })
+
+  type Interceptor = PartialFunction[Any, InterceptorResult]
+
+  private sealed trait HandlerResult
+  private case object Done extends HandlerResult
+  private case object Undefined extends HandlerResult
+
+  private type Handler = Any => HandlerResult
+}
+
 /**
  * Trait implementing Receive Pipeline Pattern. Mixin this trait
  * for configuring a chain of interceptors to be applied around
  * Actor's current behavior.
  */
 trait ReceivePipeline extends Actor {
+  import ReceivePipeline._
 
-  private var pipeline: Vector[Receive ⇒ Receive] = Vector.empty
-  private var aroundCache: Option[(Receive, Receive)] = None
+  private var pipeline: Vector[Interceptor] = Vector.empty
+  private var adviceCache: Option[(Receive, Receive)] = None
 
   /**
    * Adds an inner interceptor, it will be applied lastly, near to Actor's original behavior
    * @param around A Receive decorator. Gets by parameter the next Receive in the chain
    *               and has to return a new Receive with the decoration applied
    */
-  def pipelineInner(around: Receive ⇒ Receive): Unit = {
-    pipeline :+= around
-    aroundCache = None
+  def pipelineInner(around: Interceptor): Unit = {
+    pipeline :+= withDefault(around)
+    adviceCache = None
   }
   /**
    * Adds an outer interceptor, it will be applied firstly, far from Actor's original behavior
    * @param around A Receive decorator. Gets by parameter the next Receive in the chain
    *               and has to return a new Receive with the decoration applied
    */
-  def pipelineOuter(around: Receive ⇒ Receive): Unit = {
-    pipeline +:= around
-    aroundCache = None
+  def pipelineOuter(around: Interceptor): Unit = {
+    pipeline +:= withDefault(around)
+    adviceCache = None
+  }
+
+  private def combinedAdvisor: Receive => Receive = { receive =>
+    // So that reconstructed Receive PF is undefined only when the actor's 
+    // receive is undefined for a transformed message that reaches it...     
+    val innerReceiveHandler: Handler = {
+      case msg => receive.lift(msg).map(_ => Done).getOrElse(Undefined)
+    }
+
+    val zipped = pipeline.foldRight(innerReceiveHandler) { (outerInterceptor, innerHandler) =>
+      outerInterceptor.andThen {
+        case Inner(msg)        => innerHandler(msg)
+        case HandledCompletely => Done
+      }
+    }
+
+    toReceive(zipped)
+  }
+
+  private def toReceive(handler: Handler) = new Receive {
+    // Cache the result locally to avoid evaluating potentially 
+    // side-effecting code twice in isDefinedAt/apply...
+    var resultCache = Option.empty[(Any, HandlerResult)]
```

Comment:
Good point. Will make that change.
On 4 Jul 2015 5:37 pm, "Gastón M. Tonietti" <notifications@github.com>
wrote:

> In akka-contrib/src/main/scala/akka/contrib/pattern/ReceivePipeline.scala
> <https://github.com/akka/akka/pull/17678#discussion_r33888578>:
>
> > +    }
> > +
> > +    val zipped = pipeline.foldRight(innerReceiveHandler) { (outerInterceptor, innerHandler) =>
> > +      outerInterceptor.andThen {
> > +        case Inner(msg)        => innerHandler(msg)
> > +        case HandledCompletely => Done
> > +      }
> > +    }
> > +
> > +    toReceive(zipped)
> > +  }
> > +
> > +  private def toReceive(handler: Handler) = new Receive {
> > +    // Cache the result locally to avoid evaluating potentially
> > +    // side-effecting code twice in isDefinedAt/apply...
> > +    var resultCache = Option.empty[(Any, HandlerResult)]
>
> resultCache is never cleaned, it could unnecessarily prevent the message
> to be GCed. Maybe apply should empty it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/17678/files#r33888578>.
>